File: Arduino/Tracking/Message.cpp
#include "Message.h"

Message::Message()
{
  ID = getNextID();
  Source = 1;
  Timestamp = millis();
  Type = MsgType::Unknown; 
}

int Message::getDataLength() const
{
  return 12;
}

void Message::getData(byte * buffer) const
{
	memcpy(buffer, &ID, 2);
  memcpy(buffer + 2, &Source, 1);
	memcpy(buffer + 3, &Timestamp, 4);
  memset(buffer + 7, 0, 4);
  memcpy(buffer + 11, &Type, 1);
}

uint16_t Message::_ID = 0;

uint16_t Message::getNextID()
{
  uint16_t id = _ID;

  if (_ID >= 65535)
  {
    _ID = 0;
  }
  else
  {
    _ID = _ID + 1;
  }

  return id;
}

void Message::memcpyrvs(void * buffer, const void * value, size_t n)
{
  for (int i = n, j = 0; i > 0; j++, i--)
  {
    memcpy(buffer + j, value + i, 1);
  }
}

File: GroundStation/Libraries/XBeeComm/Message.cs
using System;
using System.IO;

namespace Baerocats.XBee
{
    public abstract class Message
    {
        public enum MsgType
        {
            Unknown = 0,
            Data = 1,
            Log = 2,
            Cmd = 3,
            CmdResponse = 4
        }

        public enum MsgSource
        {
            Ground = 0,
            Rocket = 1,
            Tripod = 2
        }

        private static ushort _ID = 0;
        private static object _IDLock = new object();
        
        private static ushort GetNextID()
        {
            ushort ID;

            lock (Message._IDLock)
            {
                ID = Message._ID;

                if (Message._ID >= 65535)
                {
                    Message._ID = 0;
                }
                else
                {
                    Message._ID++;
                }
            }

            return ID;
        }

        public ushort ID;
        public MsgSource Source;
        public ulong Timestamp;
        public MsgType Type;

        public Message()
        {
            ID = GetNextID();
            Source = 0;
            Timestamp = Convert.ToUInt64(DateTime.UtcNow.Subtract(
                new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc))
                .TotalMilliseconds * 1000);
            Type = MsgType.Unknown;
        }

        public Message(ushort id, MsgSource source, ulong timestamp, MsgType type)
        {
            ID = id;
            Source = source;
            Timestamp = timestamp;
            Type = type;
        }

        public virtual int GetDataLength()
        {
            // ID (2 bytes) + Source (1 Byte) + Timestamp (8 Bytes) + Type (1 Byte)
            return 12;
        }

        public virtual void GetData(Stream stream)
        {
            
        }
    }
}